docs(manifest): Improvements to `name`, `short_name`, `start_url`

dipikabh opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

This work is part of improving the web/manifest docs.

Notable changes in this PR:

- `name` and `short_name`:
    - Added "Description" (info from spec plus some general guidelines) and "See also" sections
    - Removed reference to `dir` and `lang` because they are not implemented (https://developer.mozilla.org/en-US/docs/Web/Manifest)
    - Updated examples and added prose

- `start_url`:
   - "Description" and "Privacy considerations" sections cover info from the spec
   - Added more explanation in examples

### Motivation

To ensure all sections have sufficient explanation, all caveats from spec are covered, and the pages follow a similar template

### Additional details

- Spec links:
  - https://w3c.github.io/manifest/#name-member
  - https://w3c.github.io/manifest/#short_name-member
  - https://w3c.github.io/manifest/#start_url-member

### Related issues and pull requests

Tracking issue: https://github.com/mdn/mdn/issues/560

dipikabh committed:
updated name and short_name

dipikabh committed:
start_url wip

dipikabh committed:
start_url updates

dipikabh commented:
@hamishwillee, I'll apply whatever we decide in the `icons` PR (for syntax and keys) here.

hamishwillee commented:
Cool. I might need another ping after you do that.

dipikabh committed:
update syntax blocks

dipikabh committed:
update keys headings

dipikabh commented:
Hi @hamishwillee, I've updated the Syntax blocks as per the new format.

hamishwillee commented on files/en-us/web/manifest/name/index.md:
```diff
@@ -7,31 +7,57 @@ browser-compat: html.manifest.name
-    </tr>
-  </tbody>
-</table>
+The `name` manifest member is used to specify the name of your web application as it's typically displayed to users.
```
Loss of information
```suggestion
The `name` manifest member is used to specify the name of your web application as it's usually displayed to users, such as in application lists or as a label for an icon.
```

dipikabh replied:
Sure, can add. Tweaked the last bit to "or as a label for your application's icon."

hamishwillee commented on files/en-us/web/manifest/name/index.md:
````diff
@@ -7,31 +7,57 @@ browser-compat: html.manifest.name
+```json-nolint
+/* Full names of web apps */
+"name": "Daily Task Planner"
+"name": "Recipe and Pantry Tracker"
````
Let's cut back to two (no real strong reason, other than I think the point has been made).
```suggestion
```

hamishwillee commented on files/en-us/web/manifest/name/index.md:
```diff
@@ -7,31 +7,57 @@ browser-compat: html.manifest.name
+> [!NOTE]
+> For Chromium-based browsers, a `name` or `short_name` is required in the manifest for the web app to be installable.
+
+Consider the following points when selecting a name for your web app:
```
Worth having a best practices heading? Just a thought, not advocating.

dipikabh replied:
Generally not a fan of a lone H3 (as an exception, we do have them in "Examples" section), but adding one here will improve the structure and readability of the information.

hamishwillee commented on files/en-us/web/manifest/name/index.md:
```diff
@@ -7,31 +7,57 @@ browser-compat: html.manifest.name
+
+Consider the following points when selecting a name for your web app:
+
+- How well it indicates the purpose or nature of your app
+- Whether it is clear and easy to understand and remember
+- Its length, especially if you're not providing a separate `short_name`
+- How it will appear in different contexts, such as in app lists or home screen
+- Its uniqueness compared to existing apps
+- Its cultural sensitivity and appropriateness for your target audience
+- How well it translates or is perceived in different languages if your app targets a global audience
+- Whether it has any trademark conflicts
```
Might be more readable to make these succinct

```suggestion
- How well it indicates the purpose or nature of your app
- Whether it is clear and easy to understand and remember
- Length of the name, especially if you're not providing a separate `short_name`
- How it will appear in different contexts, such as in app lists or home screen
- How easily it can be differentiated from other similar apps
- Cultural sensitivity and appropriateness for your target audience
- How well it translates or is perceived in different languages if your app target is a global audience
- Potential trademark conflicts
```

dipikabh replied:
These are good, incorporating. Also changing the order slightly so that the length consideration is stated upfront:

```
- Length of the name, especially if you're not providing a separate `short_name`
- How well it indicates the purpose or nature of your app
- Whether it is clear and easy to understand and remember
- How it appears in different contexts, such as in app lists or home screen
- How easily it can be differentiated from other similar apps
- Cultural sensitivity and appropriateness for your target audience
- How well it translates or is perceived in different languages if your app target is a global audience
- Potential trademark conflicts
```

hamishwillee commented on files/en-us/web/manifest/name/index.md:
```diff
@@ -7,31 +7,57 @@ browser-compat: html.manifest.name
-Simple `name` in left-to-right language:
+### Adding a name for your web app
+
+For a web app that helps users navigate trails and plan their hiking adventures, you can add the following `name` to the app manifest:
```
Since this is "for example":

```suggestion
For a web app that helps users navigate trails and plan their hiking adventures, you might add the following `name` to the app manifest:
```

hamishwillee commented on files/en-us/web/manifest/short_name/index.md:
````diff
@@ -7,39 +7,58 @@ browser-compat: html.manifest.short_name
+```json-nolint
+/* Short names of web apps */
+"short_name": "TaskPlanner"
+"short_name": "RecipePantry"
````
IFF you cut the `name` down to 2 options
```suggestion
```

hamishwillee commented on files/en-us/web/manifest/short_name/index.md:
```diff
@@ -7,39 +7,58 @@ browser-compat: html.manifest.short_name
+Keep the following points in mind when selecting a short name for your web app:
 
-`short_name` in Arabic, which will be displayed right-to-left:
+- It should reflect a concise version of your app's `name`.
```
```suggestion
- It should be a concise version of your app's `name`.
```

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,42 +7,100 @@ browser-compat: html.manifest.start_url
+  - : A string that represents the starting URL of a web app.
+    The value must be same-origin with the page that links to the manifest file.
+    If the value is relative, it is resolved against the manifest file's URL.
+    If `start_url` is unspecified or the value is invalid (i.e., not a string, not a valid URL, or not same-origin with the linking page), the URL of the page that links to the manifest file is used as a fallback.
```
"the URL of the page that links to the manifest file is used as a fallback." - only one page can link to the manifest file?

hamishwillee replied:
If so, is there a link we can provide to this info.

dipikabh replied:
I think I might have got the fallback behavior wrong. I am trying to get some clarification in https://github.com/w3c/manifest/issues/1147.

dipikabh replied:
Until the confusion about "document URL" as used in spec is cleared, I'm reverting to the previous version but with a slight change. Instead of "same-origin with page that links to manifest file", I now have "same-origin with the manifest URL". This is more plausible. 

> only one page can link to the manifest file?

Multiple pages in the app can link to the manifest file. It seems that "URL of the page that links to the manifest file" will eventually be the page from which the app is installed but I guess we'll need to wait for a response to my question in https://github.com/w3c/manifest/issues/1147.

fyi, the info was originally added via https://github.com/mdn/content/pull/34638/files but there is no explanation about the source.

dipikabh replied:
Updated the Description and Examples accordingly

hamishwillee replied:
That makes sense to me.

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,42 +7,100 @@ browser-compat: html.manifest.start_url
+
+### Values
+
+- `start_url`
```
```suggestion
- `start_url` {{optional_inline}}
```

dipikabh replied:
All manifest members are optional: https://developer.mozilla.org/en-US/docs/Web/Manifest#members.
Do you think adding the optional tag to all would make this more explicit/clear?

hamishwillee replied:
Not certain. It is odd to add them in cases like this where the value is optional, in particular when the top level keys are optional. On the other hand, there are cases where the key is optional, but it is an object that has mandatory fields if the option is defined.

If it were me I would NOT add them for these cases.

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,42 +7,100 @@ browser-compat: html.manifest.start_url
 
 > [!NOTE]
-> The `start_url` member is purely advisory, and a user agent may ignore it or allow the user to alter it at install time or afterwards.
+> For Chromium-based browsers, a `start_url` is required in the manifest for the web app to be installable.
```
We have these notes everywhere. IMO this is compatibility information that should be moved to the BCD and removed from these docs (once that happens).

dipikabh replied:
I've opened https://github.com/mdn/browser-compat-data/pull/24700

hamishwillee replied:
Thanks, so would we remove this text now or wait until the BCD merges?

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,42 +7,100 @@ browser-compat: html.manifest.start_url
+
+The `start_url` value above is valid because it is same-origin with `https://example.com/index.html`.
+
+In contrast, the following `start_url` will be considered invalid:
```
```suggestion
In contrast, the following `start_url` is invalid:
```

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
````diff
@@ -7,42 +7,100 @@ browser-compat: html.manifest.start_url
 ```
 
-### Relative URL
+In this case, the app launch will fallback to using `https://example.com/index.html`.
````
```suggestion
```

dipikabh committed:
fix feedback wip

dipikabh committed:
fix start_url value interpretation

dipikabh committed:
update glossary app context

dipikabh committed:
add meaning of applied manifest

hamishwillee commented on files/en-us/web/manifest/name/index.md:
```diff
@@ -7,31 +7,58 @@ browser-compat: html.manifest.name
+
+## Description
+
+The `name` manifest member serves as the {{Glossary("Accessible_name", "accessible name")}} for your installed app. It is displayed to users in various contexts, such as in a list of other installed web apps, as a label for your app's icon, or in the application switcher or task manager. In space-constrained contexts, such as on a device's home screen or in the application switcher, where the ful ...
```
The short name is a different "context" so let's put in own paragraph.

```suggestion
The `name` manifest member serves as the {{Glossary("Accessible_name", "accessible name")}} for your installed app. It is displayed to users in various contexts, such as in a list of other installed web apps, as a label for your app's icon, or in the application switcher or task manager.

In space-constrained contexts, such as on a device's home screen or in the application switcher, where the full `name` may not fit, the [`short_name`](/en-US/docs/Web/Manifest/short_name) value may be used instead, if available.
```

hamishwillee committed:
Update files/en-us/web/manifest/name/index.md

hamishwillee commented on files/en-us/web/manifest/name/index.md:
```diff
@@ -7,31 +7,60 @@ browser-compat: html.manifest.name
+
+In space-constrained contexts where the full `name` may not fit, such as on a device's home screen or in the application switcher, the value of the [`short_name`](/en-US/docs/Web/Manifest/short_name) property (if defined) may be used instead.
+
+> [!NOTE]
+> For Chromium-based browsers, a `name` or `short_name` is required in the manifest for the web app to be installable.
```
Are you planning on removing these before merge? Please link the BCD PR.

dipikabh replied:
Done here: https://github.com/mdn/browser-compat-data/pull/24700

hamishwillee commented on files/en-us/glossary/application_context/index.md:
```diff
@@ -1,18 +1,23 @@
+The manifest is applied after the application context is created but before navigation begins to either a start URL or a deep link. A **deep link** is a URL that directs users to a specific page within the web app, bypassing the home page.
 
-Please note that the start URL is not necessarily the value of the start_url member: the user or user agent could have changed it when the application was added to home-screen or otherwise bookmarked.
+When an application context is created, browsers must immediately navigate to a start URL or a deep link.
+This navigation replaces the current entry in the browsing history.
+If the application context is created to navigate to a deep link, the browser navigates directly to that deep link; otherwise, it navigates to the start URL.
```
FWIW this is interesting.

hamishwillee commented on files/en-us/glossary/application_context/index.md:
```diff
@@ -1,18 +1,23 @@
 {{GlossarySidebar}}
 
-An **application context** is a top-level {{glossary("browsing context")}} that has a [manifest](/en-US/docs/Web/Manifest) applied to it.
+An **application context** is a top-level {{glossary("browsing context")}} to which a [web application manifest](/en-US/docs/Web/Manifest) is applied. The members of an applied manifest affect the presentation or behavior of the browsing context.
```
This has lots of useful information. It feels a bit like it should be in a web APP guide though, rather than in a glossary.

But anyhow, as a glossary it needs perhaps a bit more context about web apps - its a quick jump into manifest and you might not know that a manifest belongs to a web app. So maybe something like...

```suggestion
**Application context** is the term used to refer to the top-level {{glossary("browsing context")}} of a [web application](/en-US/docs/Web/Progressive_web_apps).

The application context is defined by the [web application manifest](/en-US/docs/Web/Manifest), which determines how the browsing context is presented and behaves.
The manifest is applied after the application context is created but before navigation begins to either a start URL or a deep link.
A **deep link** is a URL that directs users to a specific page within the web app, bypassing the home page.
```

hamishwillee replied:
And if so, then tidying the following line.

dipikabh replied:
There is a lot of interesting info that I am discovering reading the spec and that I want to include on the landing page or in a guide (making a note of these, but doesn't look like I'll get to this exercise this year).

Great suggestions to add more clarity and context, thanks.
- I haven't used the PWA link here though because no all web apps are PWAs so that could be misleading.
- I changed "The application context is defined..." to active voice.
- I also added a definition kind-of line for start url so that it looks balanced.

> **Application context** refers to the top-level {{glossary("browsing context")}} of a web application.
> It determines how an app's browsing context, such as a tab or a window, is presented and behaves.

> Web developers define the application context in the [web app's manifest file](/en-US/docs/Web/Manifest).
> They use the [`scope`](/en-US/docs/Web/Manifest/scope) member in the manifest to specify the set of URLs that are considered part of the application context and to which the manifest applies.

> The manifest is applied after the application context is created but before navigation begins to either a start URL or a deep link.
> A **start URL** is the initial page of the web app.
> A **deep link** is a URL that directs users to a specific page within the web app, bypassing the home page.
The application context ensures that the app's defined behavior and presentation are maintained within its scope.

hamishwillee replied:
Looks good. Your reasoning makes sense for all.

hamishwillee reviewed (commented):
Looks good.
- A suggestion https://github.com/mdn/content/pull/35847/files#r1796196688
- Do you plan on removing the chromium comments to BCD first, or should we merge and do as a post process?

dipikabh committed:
apply feedback

dipikabh commented:
Thanks Hamish. This took an unexpected turn with `start_url` and I've gone back and on forth on its content (with clarifications yet to come). Could you please read through "Description" and "Examples" again because I reverted the "Values" section.

Thanks a lot for reviewing the glossary page that I added later in the cycle. I've opened a BCD PR now but we need not hold up landing this.

hamishwillee commented on files/en-us/glossary/application_context/index.md:
```diff
@@ -1,18 +1,30 @@
 {{GlossarySidebar}}
 
-An **application context** is a top-level {{glossary("browsing context")}} that has a [manifest](/en-US/docs/Web/Manifest) applied to it.
+**Application context** refers to the top-level {{glossary("browsing context")}} of a web application.
```
@dipikabh Actually, not entirely sure about "I haven't used the PWA link though because not all web apps are PWAs so that could be misleading."

You're right, but the presence of a manifest is what defines a PWA, so actually PWA is more correct than "web application". It probably should be 

```suggestion
**Application context** refers to the top-level {{glossary("browsing context")}} of a {PWA][insert link].
```

dipikabh replied:
Added like you'd originally suggested:

`...of a [web application](/en-US/docs/Web/Progressive_web_apps).`

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,36 +7,100 @@ browser-compat: html.manifest.start_url
-    </tr>
-  </tbody>
-</table>
+The `start_url` manifest member is used to specify the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.
```
I wonder if we should forward reference to "Browsers have flexibility in how they handle `start_url`."

It is pretty significant that your start URL might not be used. 

Maybe (?)

```suggestion
The `start_url` manifest member is a hint to the browser for the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.
```

dipikabh replied:
It was a note in the previous version. I've now added a note and linked to the subsection (added one now) in "Description"

> The `start_url` manifest member is used to specify the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.

> [!NOTE]
> [Browsers have flexibility](#browser_flexibility) in how they the handle `start_url` and may not always use the specified value.

dipikabh replied:
Had to update the note to avoid a link right at the beginning. `[!NOTE]` and `[` of the link right after seemed to be conflicting.

> [!NOTE]
> The `start_url` is a hint for browsers. [Browsers have flexibility](#browser_flexibility) in how they the handle `start_url` and may not always use the specified value.

hamishwillee commented:
OK, added a few minor comments. 

W.r.t. 

> For security reasons, the `start_url` must be same-origin with the manifest URL.
If a non-same-origin `start_url` is specified, browsers will fallback to using the page that links to the manifest as the default starting page.

That's probably correct. My understanding is that usually you'd have the manifest in a folder, and link it from a html file "index.html" in the same folder. YOu can be pretty sure that whatever page you have in that folder is likely to be your intended start page.

dipikabh committed:
remove chromium notes

dipikabh committed:
fix feedback

dipikabh committed:
update note

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,36 +7,103 @@ browser-compat: html.manifest.start_url
+
+    If `start_url` is unspecified or the value is invalid (i.e., not a string, not a valid URL, or not same-origin with the manifest URL), the URL of the page that links to the manifest is used.
+
+## Description
+
+### Best practices
+
+To ensure the same entry point to your app for all users, specify a `start_url`.
+This URL should navigate users to an important page of your app, such as a dashboard.
+Consider features that users would want to access immediately after launching the app.
+If your app's main page is at the root of your site, you can set the `start_url` to `/`.
+You can also specify a deep link (e.g., `https://myapp.com/product/whatsnew`) to direct users to specific content within your app.
+Avoid specifying a generic starting page.
+
+For security reasons, the `start_url` must be same-origin with the manifest URL.
+If a non-same-origin `start_url` is specified, browsers will fallback to using the page that links to the manifest as the default starting page.
+
+### Browser flexibility
+
+Browsers may ignore the specified `start_url` value or provide users with a choice not to use it.
+Browsers may allow users to modify the URL, such as when creating a bookmark for the web app or at any later time.
+As a result, the `start_url` you specify may not be the one ultimately used.
+It is important for you to keep this in mind while designing your app to allow for variations in `start_url`.
```
I don't love the title "browser flexibility" and I think this info could be near the top. I'm going to do it like this and then merge:

```suggestion
## Description

The `start_url` allows you to recommend an appropriate common entry point for all users.

Note however that browsers are not required to use the specified URL.
They might ignore the specified value or provide users with alternative choices.
The might also allow users to modify the URL at a later time, such as when creating a bookmark for the web app.
Keep this in mind when designing your app .

### Best practices

This URL should navigate users to an important page of your app, such as a dashboard.
Consider features that users would want to access immediately after launching the app.
If your app's main page is at the root of your site, you can set the `start_url` to `/`.
You can also specify a deep link (e.g., `https://myapp.com/product/whatsnew`) to direct users to specific content within your app.
Avoid specifying a generic starting page.

For security reasons, the `start_url` must be same-origin with the manifest URL.
If a non-same-origin `start_url` is specified, browsers will fallback to using the page that links to the manifest as the default starting page.
```

hamishwillee commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,36 +7,103 @@ browser-compat: html.manifest.start_url
 
-The `start_url` member is a string that represents the _start URL of the web application_ — the preferred URL that should be loaded when the user launches the web application (e.g., when the user taps on the web application's icon from a device's application menu or homescreen).
+> [!NOTE]
+> The `start_url` is a hint for browsers. [Browsers have flexibility](#browser_flexibility) in how they the handle `start_url` and may not always use the specified value.
```
```suggestion
> The `start_url` is a hint for browsers. [Browsers have flexibility](#description) in how they the handle `start_url` and may not always use the specified value.
```

hamishwillee committed:
Improve description and browser flexibility link

hamishwillee reviewed (approved):
G'day @dipikabh 

Made a few minor changes and am going to merge. If you aren't entirely happy with those we can fix as a post process. Pretty happy with this all thanks.

dipikabh commented on files/en-us/web/manifest/start_url/index.md:
```diff
@@ -7,36 +7,102 @@ browser-compat: html.manifest.start_url
+
+The `start_url` allows you to recommend an appropriate common entry point for all users.
+
+Note however that browsers are not required to use the specified URL.
+They might ignore the specified value or provide users with alternative choices.
+The might also allow users to modify the URL at a later time, such as when creating a bookmark for the web app.
+Keep this in mind when designing your app .
```
- Fixing few typos plus keeping some text similar to the spec.
- From the spec: "The [start_url](https://w3c.github.io/manifest/#dfn-start_url) member is purely advisory, and a user agent MAY [ignore](https://w3c.github.io/manifest/#dfn-ignore) it or provide the end-user the choice not to make use of it. A user agent MAY also allow the end-user to modify the URL when, for instance, a bookmark for the web application is being created or any time thereafter."
```suggestion
Note, however, that browsers are not required to use the specified URL.
They may ignore the specified value or provide users with a choice not to use it.
They may also allow users to modify the URL when creating a bookmark for the web app or at a later time.
Keep this in mind when designing your app to allow for variations in `start_url`.
```

dipikabh committed:
Update files/en-us/web/manifest/start_url/index.md

dipikabh commented:
Thanks, @hamishwillee. I fixed a few nits and reverted some text so that it is closer to what the spec suggests.

I'll merge this now.
